tests/test_spend.py

```diff
@@ -252,17 +252,17 @@ def test_send_to_self(lianad, bitcoind):
     lianad.rpc.broadcastspend(spend_txid)
 
     # The only output is the change output so the feerate of the transaction must
-    # not be lower than the one provided, and only possibly slightly higher (since
-    # we slightly overestimate the satisfaction size).
-    # FIXME: a 15% increase is huge.
+    # not be much lower than the one provided (it could be slightly lower since
+    # the change amount is determined using feerate in terms of sat/wu which, due
+    # to rounding, can lead to a slightly lower feerate in terms of sat/vb,
+    # especially when the number of inputs increases), and only possibly slightly
+    # higher (since we slightly overestimate the satisfaction size).
     res = bitcoind.rpc.getmempoolentry(spend_txid)
-    spend_feerate = int(res["fees"]["base"] * COIN / res["vsize"])
-    if not USE_TAPROOT:
-        assert specified_feerate <= spend_feerate <= int(specified_feerate * 115 / 100)
+    spend_feerate = res["fees"]["base"] * COIN / res["vsize"]  # keep as decimal
+    if USE_TAPROOT:
+        assert specified_feerate <= spend_feerate < specified_feerate + 0.5
     else:
-        # FIXME: under Taproot we should not consider the max feerate of all leaves if there
-        # is a spendable internal key.
-        assert specified_feerate <= spend_feerate <= int(specified_feerate * 125 / 100)
+        assert specified_feerate - 0.5 < spend_feerate < specified_feerate + 0.5
```

Comment:
Why is there still this discrepancy between Taproot and P2WSH? That we could now get a spend feerate lower than the specified feerate is also concerning.

Reply:
I found the actual P2WSH feerate for the same test can be either above or below the target feerate. Perhaps the higher values occur for low-r ECDSA signatures that are smaller than the max satisfaction size.

In this particular instance, I found the Taproot feerate was always above the target feerate and so kept two separate assertions, but I think in general it can also be smaller than the target feerate.

Reply:
How comes? I thought we were always overestimating? Underestimating is more serious and should never happen.

> Perhaps the higher values occur for low-r ECDSA signatures that are smaller than the max satisfaction size.

This could explain the higher value (since we assume signature are always high-r in the estimation) but not the lower one.

Reply:
> Underestimating is more serious and should never happen.

It's due to rounding up when converting from weight to vb. For example, if we have a target feerate of 8 sat/vb, the coin selection will use 2 sat/wu as the target. If it finds a transaction with weight, say, 1003 wu, then it will pay a fee of at least 2006 sats. The weight in vb is 251, which gives a feerate of at least 7.99 sat/vb.

Note that if the target feerate is 1 sat/vb, the resulting feerate will not be smaller. For 1 sat/vb, this gives a target of 0.25 sat/wu, and the fee is at least `ceil(0.25 * weight)` sats, but this is also the size of the transaction in vb.

We might be able to modify how coin selection is done to ensure we get the target feerate in sat/vb, e.g. by iteratively increasing the target sat/wu until we reach the target sat/vb.

Reply:
> We might be able to modify how coin selection is done to ensure we get the target feerate in sat/vb, e.g. by iteratively increasing the target sat/wu until we reach the target sat/vb.

It might be possible to do this via the `score()` method of our selection metric so that instead of returning the fee based on the sat/wu target, it returns the fee based on the sat/vb target, which would avoid the need to iterate over sat/wu values.

Reply:
But even better if there were an option in bdk_coin_select to target the feerate in sat/vb instead of internally using sat/wu.

Reply:
Alright, so this is orthogonal to this PR. Can we somehow also round up when converting from sat/vb to sat/wu to counteract this effect? I'd rather we always very slightly overpay than sometimes create transaction which can't propagate..

Reply:
I don't think we can a priori because how the value is rounded up will depend on the weight of the final transaction.

From a quick look, it should be fairly straightforward to add this option to the coin-select crate so that if you specify the feerate in sat/vb, it calculates the fee based on the transaction's vsize instead of weight. In the shorter term, we could add this logic on our side only to the modified metric we're using.

Reply:
I see that mempool.space calculates the vsize as weight/4, without rounding up, and uses this for their sat/vb feerate calculation. This would be another option, but I don't think it's technically correct.

Reply:
We should track the potential underestimation in a separate issue and try to come up with a fix for v7.

Reply:
For reference, https://github.com/wizardsardine/liana/issues/1132 was created.